MozziGuts_impl_ARDUINOFSP.hpp

```diff
@@ -0,0 +1,227 @@
+void dac_creation(pin_size_t pinN) {
+  if (IS_DAC(pinN)) {
+    auto cfg_dac = getPinCfgs(pinN, PIN_CFG_REQ_DAC);
+    pin = cfg_dac[0];
+    if (IS_DAC_8BIT(pin)) {
+      dac_bits = 8;
+      dtc_info.transfer_settings_word_b.size = TRANSFER_SIZE_1_BYTE;
+      if (GET_CHANNEL(pin) == 0) {
+#ifdef DAC_ADDRESS_8_CH0
+        dtc_info.p_dest = (void *)DAC_ADDRESS_8_CH0;
+#endif
+      } else if (GET_CHANNEL(pin) == 1) {
+#ifdef DAC_ADDRESS_8_CH1
+        dtc_info.p_dest = (void *)DAC_ADDRESS_8_CH1;
+#endif
+      }
+    } else {
+      dac_bits = 12;
+      dtc_info.transfer_settings_word_b.size = TRANSFER_SIZE_2_BYTE;
+      if (GET_CHANNEL(pin) == 0) {
+#ifdef DAC_ADDRESS_12_CH0
+        dtc_info.p_dest = (void *)DAC_ADDRESS_12_CH0;
+#endif
+      } else if (GET_CHANNEL(pin) == 1) {
+#ifdef DAC_ADDRESS_12_CH1
+        dtc_info.p_dest = (void *)DAC_ADDRESS_12_CH1;
+#endif
+      }
+    }
+  }
+
+  dtc_info.transfer_settings_word_b.dest_addr_mode = TRANSFER_ADDR_MODE_FIXED;
+  dtc_info.transfer_settings_word_b.repeat_area = TRANSFER_REPEAT_AREA_SOURCE;
+  dtc_info.transfer_settings_word_b.irq = TRANSFER_IRQ_END;
+  dtc_info.transfer_settings_word_b.chain_mode = TRANSFER_CHAIN_MODE_DISABLED;
+  dtc_info.transfer_settings_word_b.src_addr_mode =
+    TRANSFER_ADDR_MODE_INCREMENTED;
+  dtc_info.transfer_settings_word_b.mode = TRANSFER_MODE_REPEAT;
+  dtc_info.p_src = (void const *)NULL;
+
+  dtc_info.num_blocks = 0;
+  dtc_info.length = 0;
+
+  dtc_cfg.p_info = &dtc_info;
+  dtc_cfg.p_extend = &dtc_cfg_extend;
+
+  dtc_cfg_extend.activation_source = FSP_INVALID_VECTOR;
+}
+
+void dac_init() {
+  if (IS_DAC_8BIT(pin)) {
+#if DAC8_HOWMANY > 0
+    if (GET_CHANNEL(pin) < DAC8_HOWMANY) {
+      _dac8[GET_CHANNEL(pin)].init();
+    }
+#endif
+  } else {
+    if (GET_CHANNEL(pin) < DAC12_HOWMANY) {
+      _dac12[GET_CHANNEL(pin)].init();
+    }
+  }
+}
+
+#endif
+
+
+//////////////// TIMER ////////////////
+
+#if EXTERNAL_AUDIO_OUTPUT == true
+void timer_callback_dummy(timer_callback_args_t __attribute__((unused)) *args){defaultAudioOutput();};
+#else
+//void timer_callback_dummy(timer_callback_args_t __attribute__((unused)) *args){
+void timer_callback_dummy(timer_callback_args_t __attribute__((unused)) *args){output_buffer.read();}; // to empty the buffer (the dac does not take care of it), a bit a waste of timer...
+#endif
+
+void timer_init() {
+  uint8_t type = 0;
+  int8_t tindex = FspTimer::get_available_timer(type);
+
+  if (tindex < 0) {
+    tindex = FspTimer::get_available_timer(type, true);
+  }
+
+  if (tindex >= 0) {
+    timer.begin(TIMER_MODE_PERIODIC, type, tindex, AUDIO_RATE, 50.0,timer_callback_dummy);
+    timer.setup_overflow_irq();
+  }
+  
+#if EXTERNAL_AUDIO_OUTPUT != true // we need to set up another timer for dac caring
+  // note: it is running at the same speed than the other one, but could not manage
+  // to get the other one updating the dac and removing the samples from the buffer…
```

Comment:
Why not? If you simply add timer_overflow_dummy as a callback timer_dac, shouldn't that be just enough?

Reply:
I left this one for either when I have a good chunk of uninterrupted (pun intended) time before me, or someone else. I remember trying quite hard to make that work with one timer, to not avail. Getting back into the more or less document source files of these timers gets some time.

Reply:
(From what I remember, adding the overflow_dummy as a callback_timer_dac would mess up with the DAC, which would not get updated anymore)